README
Xmp Manager is a Gui to write xmp metadata on .png, .jpeg files.


more info on: http://grigio.org/xmp_manager



debian/changelog
xmp-manager (0.3.2-0ubuntu1) jaunty; urgency=low

  * Integrated DigiWf with Nautilus Actions so that the shortcuts are created
    automatically upon installation. Right-click on a directory to "File 
    images". Right-click on picture (or a selection of multiple pictures) to 
    "Edit metadata..."
  * Removed unused UI elements.
  * Added an icon to the GUI.

 -- Stefaan Vanderheyden <mrsvan@hotmail.com>  Fri, 24 Apr 2009 23:05:49 +0200

xmp-manager (0.3.1-0ubuntu1) jaunty; urgency=low

  * Initial Release.

 -- Stefaan Vanderheyden <mrsvan@hotmail.com>  Wed, 22 Apr 2009 10:43:06 +0200


spec/main_spec.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

require './spec/spec_helper.rb'

module XmpManager

describe File do
  before(:each) do
    @gnome_cup = XmpManager::File.new('spec/images/gnome-cup.jpg')
    @ubuntu_cup = XmpManager::File.new('spec/images/ubuntu-cup.jpg')
  end

  it "should load fields" do
    @gnome_cup.field['creator'].should == 'luigi'
    @gnome_cup.field['title'].should == 'un titolò'
  end


  it "should read tags" do
    @gnome_cup.tags.should == ['cup', 'gnome']
  end
  
  it "should be able to write a field" do
    @ubuntu_cup.field['title'] = 'my cup2'
    #@ubuntu_cup.save
    #@ubuntu_cup.reload!
    @ubuntu_cup.field['title'].should == 'my cup2'
    # restore
    @ubuntu_cup.field['title'] = 'my cup1'
    #@ubuntu_cup.save
  end
end

describe Selection do
  before(:each) do
    @selection = XmpManager::Selection.new(['spec/images/gnome-cup.jpg', 'spec/images/ubuntu-cup.jpg'])
  end

  it "should manage field" do
    @selection.field('title').should  == 'ubucup'
    
    # dynamic methods
    @selection.title_changed?.should == nil # INCONSISTENT
    @selection.title = 'ubucup'
    @selection.title_changed?.should == true
    @selection.title = '' # or nil
    @selection.title_changed?.should == false
    
    # fake write
  end

  it "should manage tags" do
    
    # With releated tags
    #@selection.tags.should == ["cup", "gnome", "logo", "tux", "ubuntu"]
    
    @selection.tags.should ==["cup", "gnome", "ubuntu"]
    @selection.tag_status('cup').should == true
    @selection.tag_status('tux').should == nil
    
    @selection.tag_del('cup')
    @selection.tag_status('cup').should == false
    
    @selection.tag_add('bau')
    @selection.tag_status('bau').should == true
    
  end

  # TODO remove sort
#  it "should load related tags" do
#    @selection.tags.should == ["cup", "gnome", "logo", "tux", "ubuntu"]
#  end

#  it "should load related tags with values" do
#    @selection.tags['cup'].should == ENABLED
#    @selection.tags['a non existent tag'].should == nil
#  end

end

end


spec/spec_helper.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

require 'stringio'
 
dir = File.dirname(__FILE__)
lib_path = File.expand_path("#{dir}/../lib")
$LOAD_PATH.unshift lib_path unless $LOAD_PATH.include?(lib_path)
$_spec_spec = true # Prevents Kernel.exit in various places
 
require 'spec'
require 'spec/mocks'
spec_classes_path = File.expand_path("#{dir}/../spec/spec/spec_classes")

DEBUG = false
#require 'xmpmanager/file'
require 'xmpmanager/selection'
#require 'xmpmanager/backend/exiftool'

module Spec
  module Example
    class NonStandardError < Exception; end
  end
 
  module Matchers
    def fail
      raise_error(Spec::Expectations::ExpectationNotMetError)
    end
 
    def fail_with(message)
      raise_error(Spec::Expectations::ExpectationNotMetError, message)
    end
 
    def exception_from(&block)
      exception = nil
      begin
        yield
      rescue StandardError => e
        exception = e
      end
      exception
    end
    
    def run_with(options)
      ::Spec::Runner::CommandLine.run(options)
    end
 
    def with_ruby(version)
      yield if RUBY_PLATFORM =~ Regexp.compile("^#{version}")
    end
  end
end
 
def with_sandboxed_options
  attr_reader :options
  
  before(:each) do
    @original_rspec_options = ::Spec::Runner.options
    ::Spec::Runner.use(@options = ::Spec::Runner::Options.new(StringIO.new, StringIO.new))
  end
 
  after(:each) do
    ::Spec::Runner.use(@original_rspec_options)
  end
  
  yield
end
 
def with_sandboxed_config
  attr_reader :config
  
  before(:each) do
    @config = ::Spec::Example::Configuration.new
    @original_configuration = ::Spec::Runner.configuration
    spec_configuration = @config
    ::Spec::Runner.instance_eval {@configuration = spec_configuration}
  end
  
  after(:each) do
    original_configuration = @original_configuration
    ::Spec::Runner.instance_eval {@configuration = original_configuration}
    ::Spec::Example::ExampleGroupFactory.reset
  end
  
  yield
end


xmpmanager.rb
#!/usr/bin/ruby

# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <riccio_@t_inmail_sk>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

$LOAD_PATH << File.expand_path(File.dirname(__FILE__)+'/lib')
DATA_DIR = File.expand_path(File.dirname(__FILE__))+'/data'
EXIFTOOL = "/usr/bin/exiftool"
DEBUG = false

require 'monitor'
require 'libglade2'

#
# Just to debug better
#
class Object
  def method_name
    if  /`(.*)'/.match(caller.first)
      return $1
    end
    nil
  end
end

class String
  def prepare_for_shell
    self.gsub(/\s/,'\ ')
  end
end

#
# It executes really the actions
#
class Controller
  def initialize
    @metadata = Hash.new
    load_file_paths #set @files if possible
    load_metadata
    chunks = @files[0].split('/')
    @current_directory = chunks[0..chunks.size-2].join('/') || "."
    #View.new(self)
  end
  
  def valid_tags_cache?(filename)
    puts "\n\n filename:#{filename}"
    begin
    if File::mtime(filename).min == Time.now.min
      return true
    end
    rescue
    end
      return false
  end
  
  #
  # Return the tags of other photos
  #
  def external_tags
    all_tags=[]
    filename = "#{@current_directory}/.tags.cache"
    
    # a recent cache    
    if valid_tags_cache?(filename)
      File.open(filename).each_line{|tag| all_tags << tag.strip}
    else    
    # scan the filesystem
    	cmd = open("|#{EXIFTOOL} #{@current_directory.prepare_for_shell} -xmp:subject")
		  while (temp = cmd.gets)
		    # good string with :
    		if temp[0..32] == 'Subject                         :'
      		#temp = cmd.gets
      		key, value = parse_line(temp)
      		all_tags += value.split(', ') unless value.nil?
      	end
    	end
		  cmd.close
		  puts "\n>> #{method_name}\n"+all_tags.inspect if DEBUG
    end

		ext_tags = all_tags.uniq - @tags
		ext_tags ||= []
		# HACK
		@external_tags = ext_tags
		return ext_tags
  end
  
  def tags
		cmd = open("|#{EXIFTOOL} #{@files[0].prepare_for_shell} -xmp:subject")
		#while (temp = cmd.gets)
		temp = cmd.gets
			key, value = parse_line(temp)
			@tags = value.split(', ') unless value.nil?
		#end
		cmd.close
		puts "\n>> #{method_name}\n"+@tags.inspect if DEBUG
		@tags ||= []
		return @tags
  end
  
  def metadata(field)
    @metadata[field]
  end
  
  def write(field, value)
		# hihi
		#Thread.new {
		@files.each do |filename|
			# TODO: multiple command & errors
			command = "|#{EXIFTOOL} #{filename.prepare_for_shell} -xmp:#{field}=\"#{value}\" -overwrite_original_in_place"
			cmd = open(command)
			puts "\n>> #{method_name}\n"+command if DEBUG
			#while (temp = cmd.gets)
			cmd.close
		end
		#}
		puts "\n>> #{method_name}\n"+field+"-"+value if DEBUG
	end
  
  def write_tags
    @files.each do |filename|
			xmp_subjects=''
			@tags.each do |tag|
			  xmp_subjects << " -xmp:subject=\"#{tag}\""
			end
			xmp_subjects = " -xmp:subject=\"\"" if xmp_subjects == ''
			command = "|#{EXIFTOOL} #{filename.prepare_for_shell} #{xmp_subjects} -overwrite_original_in_place"
			cmd = open(command)
			puts "\n>> #{method_name}\n"+command if DEBUG
			cmd.close
		end
  end
  
  def add_tag(value)
    @tags << value
    write_tags
	end
  
  def delete_tag(value)
	  @tags.delete value
	  write_tags
	end
  
  def load_file_paths(array = nil)
    # ARGV[0..-1]
    @files = array # from GUI
    @files ||= ARGV[0..-1] if ARGV
    #@files ||= ["test-img-tante/a.jpg", "test-img-tante/b.jpg"]
    if @files.empty?
      puts "\n>> #{method_name}\n"+"You need to specify at least a file!"
      exit(1)
    end
    puts "\n>>  #{method_name}\n"+@files.inspect+"  \n end\n" if DEBUG
  end

  
  # initially only field from the first selection are checked
  def load_metadata
    cmd = open("|#{EXIFTOOL} #{@files[0].prepare_for_shell} -xmp:all")
	  while (temp = cmd.gets)
		  key, value = parse_line(temp)
		  @metadata[key] = value unless key.nil?
	  end
	  cmd.close
	  puts "\n>> #{method_name}\n"+@metadata.inspect+"\n END \n" if DEBUG
  end
  
  def quit
    #save external_tags
    File::open("#{@current_directory}/.tags.cache",'w'){|f| f.write(@external_tags.join("\n"))}    
  end
  
	def parse_line(line)
		# "Country                         : italia".downcase.strip
		key, value = line.split(': ') unless line.nil?
		return nil, nil if key.nil? || value.nil?
		key.downcase!.gsub!(' ','')
		value[-1]='' #remove a capo
		return key, value
	end
  
end

class View

  def initialize
    @controller = Controller.new
    init_gui
    setup_tag_box
    update_gui
    
    @lock= Monitor.new
		@confirm_secs= 1
		@ready_to_write= true
		
		@metadata_dialog.show_all
		
  end
  
  
  #
  # It fills the widgets
  #
  def update_gui
    # custom fields
    @glade.widget_names.each do |name|
    	begin
      	if name[-5..-1]=='_auto' #auto event widget
          
          @glade[name].text = @controller.metadata(name.gsub('_auto', ''))

      	end
      rescue
    	end
    end
  end
  
  #
  # It loads the widgets
  #
  def init_gui
    @glade = GladeXML.new(DATA_DIR+"/nautilus-xmp-manager.glade") {|handler| method(handler)}

    # Widgets used not autogenerated
    @tag_box = @glade['tag_box']
    @tag_entry = @glade['tag_entry']
    @add_tag_button = @glade['add_tag_button']


    # Autogenerated variables and by the entry-name_auto
    @glade.widget_names.each do |name|
    	begin
      	instance_variable_set("@#{name.gsub('_auto', '')}".intern, @glade[name])
      	if name[-5..-1]=='_auto'
          name = name.gsub('_auto', '')
          
  				eval <<-END
						@#{name}.signal_connect('changed') do
						 #puts @#{name}.text
						 try_to_write(@#{name})
						end
  				END
  				
      	end
      rescue
    	end
    end

    # Widget signals
    @metadata_dialog.signal_connect('destroy'){
    	quit
    }
    @close_button.signal_connect('pressed'){
    	quit
    }
    
    @add_tag_button.signal_connect('pressed'){
      update_tag_box
    }
    
    
    @glade['help_button'].signal_connect('clicked'){
      #
    }
    

    # 		for widget in @glade.widgets
    # 			if widget.class? Gtk::Textbox #or combo
    # 				#widget.add_method(:signal_connect('changed') {try to})
    # 			end
    # 		end

    
    # Widget default values
    @metadata_dialog.title='Metadata Manager'
    @notebook.page = 0 # force page 0 see glade
  end
  
  #
  # It fills the tag_box with tags and external_tags found
  #
  def setup_tag_box
    @controller.tags.each do |tag|
      create_tag_checkbutton(tag, true)
    end
    @controller.external_tags.each do |tag|
      create_tag_checkbutton(tag)
    end
  end
  
  def is_valid_tag(tag)
    puts "\n>> #{method_name}\n"+@tag_entry.text.inspect if DEBUG
    # FIXME it should permit only utf8 a-z an spaces
    tag != ''
  end
  
  
  #
  # Add a tag that is not present in tag_box from user input 
  #
  def update_tag_box
    if is_valid_tag @tag_entry.text
      checkbutton = @tag_entry.text
      create_tag_checkbutton(checkbutton, true)
      @tag_entry.text=''
    end  
  end
  
  #
  # Create the checkbutton with a value set and adds the signal to manage it
  #
  def create_tag_checkbutton(label, active = false)
    # HACK it bans some annoying characters
    # BUG àèìòù and others can't be used :(
    # checkbutton=checkbutton.gsub(':','_').gsub('=','_').gsub('.','_').gsub('-','_')
  	
  	# FIXME Each runtime cb should have a unique name
  	checkbutton = 'cb'+rand(1000).to_s
  	 
    #   	 gtk_widget_ref(widget);
    #      gtk_container_remove(GTK_CONTAINER(old_parent), widget);
    #      gtk_container_add(GTK_CONTAINER(new_parent), widget);
    #      gtk_widget_unref(widget);
  	
  	eval <<-END
     @#{checkbutton} = Gtk::CheckButton.new('#{label}')
     # FIXME WARNING to fix
     @tag_box.pack_start(@#{checkbutton}, false, false)
     @#{checkbutton}.active = active
     @#{checkbutton}.signal_connect('toggled') do
      if @#{checkbutton}.active?
        @controller.add_tag @#{checkbutton}.label
      else
        @controller.delete_tag @#{checkbutton}.label
      end
     end
     @tag_box.add @#{checkbutton}.show
    END
    
    
     puts "\n>> #{method_name}\n"+checkbutton.inspect if DEBUG
  end
  
  #
  # Enabled everytime a char is pressed in a texbox
  #
	def try_to_write(widget)

    	# interrompe thread di scrittura se esiste, altrimenti scritture multiple
			@tmain.kill if @tmain.is_a? Thread

    	@previous_title= widget.text

    	#Thread principale
    	@tmain= Thread.new do    	

    	  sleep @confirm_secs #conferma

    		if (@previous_title == widget.text)
    			@t1= Thread.new do
    				@lock.synchronize do
    					puts "\n>> #{method_name}\n"+
    					     "#{@previous_title} poi #{widget.text} @@ #{widget.name.gsub('_auto', '')}" if DEBUG
    					@controller.write(widget.name.gsub('_auto', ''), widget.text)
    				end
    			end
				end
			
			end #thread
			#@tmain.join
			
	end
  
  def load_file_paths
    update_gui
  end
	
	def quit
	  @controller.quit
	  # HACK: a writing couldn't be finished yet
	  while @tmain.alive? do
      sleep 0.1
    end
		Gtk.main_quit
	end

end

Gtk.init
View.new
Gtk.main

## DEBUG INPUT from Nautilus-actions 
# Gtk.init
# window = Gtk::Window.new
# button=Gtk::Button.new
# window.add button
# button.label = ARGV.inspect
# window.show_all
# Gtk.main



lib/xmpmanager/file.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

require 'xmpmanager/backend/exiftool'

module XmpManager

class File
  # backend specific methods
  include XmpManager::Exiftool

  def initialize(path)
    @path = path
    @field = load_fields(path)
  end

  def path
    @path
  end
  
  def field
    @field
  end
  alias_method :fields, :field
  
  # is selection? TODO: write tag?
  def tags
    if @field['subject']
      @field['subject'].split(', ').sort
    else
      []
    end
  end

  def reload!
    
  end

#  def save
#    
#  end
  
  
  #private
  

end

end


lib/xmpmanager/ui/gtk.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

require 'gtk2'
require 'gettext'

## Localization support
#unless defined? GetText
#  module GetText
#    module_function
#    def _(msgid); msgid; end
#    def N_(msgid); msgid; end
#    def n_(msgid, msgid_plural, n)
#      msgid
#    end
#  end
#end

module XmpManager


class MainWindow

  def _(sentence)
    GetText.gettext(sentence)
  end
  
  def initialize  
    GetText.bindtextdomain('xmpmanager', PO_DIR)
    GetText.textdomain("xmpmanager")

    # DEBUG
    puts _('Title:')+"\n"
    
    @selection = XmpManager::Selection.new(ARGV)
    
    init_gui
  end
  
  def init_gui
    Gtk.init
    b = Gtk::Builder.new
    b.add_from_file(DATA_DIR+'/xmpmanager/nautilus-xmp-manager.ui')
    b.translation_domain = 'xmpmanager'# TODO: select the language
    
    puts _('Title:')+"\n"
    
    # It generates dynamic methods/signals called by UI
    b.connect_signals{|name|
      method(name)
    }
    
    # It generates signals from avalaible fields in the UI
    b.objects.each do |obj|
    	begin
    	  name = obj.name
      	instance_variable_set("@#{name.gsub('_auto', '')}".intern, b.get_object(name))
      	if name[-5..-1]=='_auto'
          name = name.gsub('_auto', '')
          
  				eval <<-END
  				  @#{name}.text = @selection.#{name}
						@#{name}.signal_connect('changed') do
						 @selection.#{name} = @#{name}.text
						end
  				END
  				
      	end
      rescue
    	end
    end
    # FIXME: merge fields with default template OR template without dynamic?
    
        
    @window = b.get_object('window')
    @tag_box = b.get_object('tag_box')
    @tag_entry = b.get_object('tag_entry')
    @add_tag_button = b.get_object('add_tag_button')
    @save_button = b.get_object('save_button')
    
    b.objects.each do |obj|
      begin
        #unless obj.class? == Entry
          obj.label = _(obj.text)
        #end
      rescue
      end
    end
    
    @selection.tags.each do |tag|
      create_tag_checkbutton(tag, true)
    end
    
    
    @window.show_all
    
    Gtk.main
  end
  
  def confirm(message)
    #TODO
    true
  end
  
  ## GUI methods/signals
  
  def on_add_tag_button_pressed
	  checkbutton = @tag_entry.text
    create_tag_checkbutton(checkbutton, true)
	  @tag_entry.text='' # valid
	  # non valid exception
  end
  
  def on_save_button_clicked
    @selection.save
    # quit
  end
  
  def on_erase_button_clicked
    @selection.erase
  end
  
  def on_window_destroy
    #@selection.changed? Gtk.main_quit : confirm('The changes made will be lost')
    Gtk.main_quit
  end
  
  def update_tag_box
    if is_valid_tag @tag_entry.text
  	  checkbutton = @tag_entry.text
      create_tag_checkbutton(checkbutton, false)
  	  @tag_entry.text=''
  	end  
  end
  
  #
  # Create the checkbutton with a value set and adds the signal to manage it
  #
  def create_tag_checkbutton(label, active = false)
    # HACK it bans some annoying characters
    # BUG àèìòù and others can't be used :(
    # checkbutton=checkbutton.gsub(':','_').gsub('=','_').gsub('.','_').gsub('-','_')
  	
  	# FIXME Each runtime cb should have a unique name
  	checkbutton = 'cb'+rand(1000).to_s
  	@selection.tag_add(label) if active
  	
    #   	 gtk_widget_ref(widget);
    #      gtk_container_remove(GTK_CONTAINER(old_parent), widget);
    #      gtk_container_add(GTK_CONTAINER(new_parent), widget);
    #      gtk_widget_unref(widget);
  	
  	eval <<-END
     @#{checkbutton} = Gtk::CheckButton.new('#{label}')
     # FIXME WARNING to fix
     @tag_box.pack_start(@#{checkbutton}, false, false)
     @#{checkbutton}.active = active
     @#{checkbutton}.signal_connect('toggled') do
      if @#{checkbutton}.active?
        @selection.tag_add @#{checkbutton}.label
      else
        @selection.tag_del @#{checkbutton}.label
      end
     end
     @tag_box.add @#{checkbutton}.show
    END
    
  end

  def quit
    Gtk.main_quit
  end

end

end


lib/xmpmanager/selection.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

require 'xmpmanager/file'

module XmpManager

class Selection
  # backend specific methods
  include XmpManager::Exiftool

  def initialize(paths)
    # clean *paths ARGV ?
    #paths = 'spec/images/write-tests/a.jpg', 'spec/images/write-tests/b.jpg' if DEBUG
    @files = load_files(paths) # return a File Array
    @field, @status, @tags = {}, {}, {}
    #@field, @status, @tags = 
    load_selection # from @files
    load_dynamic_methods
    #load_dynamic_methods
    #load_template # common fields
  end
   
  def field(key = nil)
    if key.nil?
      @field
    else
      @field[key]
    end
  end  
  alias_method :fields, :field

  # Old style method *not dynamic*
  def field_set(key, value)
      @field[key] = value
  end  
  
  # it takes the intersection of common tags
  def tags
    @tags.keys.sort
  end
  
  def tag_add(tag)
    @tags[tag] = true
  end
  
  def tag_del(tag)
    @tags[tag] = false
  end
  
  def tag_status(tag)
    @tags[tag]
  end
  
  def status(key)
    @status[key]
  end
  
  def erase
    @files.each do |file|
      file.erase
    end
  end
  
  def save
    @files.each do |file|
      @field.each do |key, value|
        file.field[key] = value
      end
      
      tags_to_write = ""
      @tags.each{|key, value| tags_to_write << key+", " if @tags[key] != false} #TODO: fix inconsistencies
      file.field['subject'] = tags_to_write
      
      file.save
    end
  end
  
  private
  
  def load_files(paths)
    files = []
    paths.each do |path| 
      files << XmpManager::File.new(path)
    end
    files
  end
  
  def load_selection
    # load default fields FIXME: move to template
    @field.merge! 'description' => '', 'creator' => '', 'rights' => '', 'title' => ''
    # load data
    @files.each do |file|
       @field.merge! file.fields
    end
    # load tags #FIXME: nil false true
    @files.each do |file|
      file.tags.each do |tag|
        @tags.merge! tag => true
      end
    end
  end

  def load_dynamic_methods
    # TODO: changed? false
    # ex. sel.title, sel.title = <value>, sel.title_changed?
    @field.each_key do |key|
      eval <<-STR
        def #{key}=(value)
          unless value.empty?
            @field['#{key}'] = value
            @status['#{key}'] = true
          else
            @status['#{key}'] = false
          end
        end
        def #{key};  @field['#{key}']; end
        def #{key}_changed?;  @status['#{key}']; end
      STR
    end
  end
  
end

end


lib/xmpmanager/backend/exiftool.rb
# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

module XmpManager

module Exiftool
  EXIFTOOL = "/usr/bin/exiftool"
  DEBUG = "true" # FIXME
  
  def load_fields(path)
    fields = {}
    command = "|#{EXIFTOOL} \"#{path}\" -xmp:all"
    puts ">> #{command}" if DEBUG
    cmd = open(command)
	  while (temp = cmd.gets)
		  key, value = parse_line(temp)
		  fields[key] = value unless key.nil?
	  end
	  cmd.close
	  fields
  end  
  
  def save
    field_list = ''
    fields.each do |field, value|
      unless field == 'subject'
        field_list << " -xmp:#{field}=\"#{value}\""
      else
        if not tags.empty?
          tags = value.split(", ")
          tags.each {|tag| field_list << " -xmp:#{field}=\"#{tag}\""}
        else
          field_list << " -xmp:subject=\"\" "
        end
      end
	  end
	  command = "#{EXIFTOOL} #{path.gsub(' ','\ ')} #{field_list} -overwrite_original_in_place"
	  puts ">> #{command}" if DEBUG
	  cmd = system command
  end
  
  def erase
	  command = "#{EXIFTOOL} #{path.gsub(' ','\ ')} -all= -overwrite_original_in_place"
	  puts ">> #{command}" if DEBUG
	  cmd = system command    
  end
  # --
  
  private
    
	def parse_line(line)
		# "Country                         : italia".downcase.strip
		key, value = line.split(': ') unless line.nil?
		return nil, nil if key.nil? || value.nil?
		key.downcase!.gsub!(' ','').gsub!("/",'_') # To generate valid methods in Selection
		value[-1]='' #remove a capo
		return key, value
	end
	
end

end


xmp-manager.rb
#!/usr/bin/ruby

# XMP Manager is a GUI to easly write XMP Metadata and tags on files
# 
# Copyright (C) 2007 Luigi Maselli <luigix_@t_gmail_com>
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
# St, Fifth Floor, Boston, MA 02110-1301 USA

$LOAD_PATH << File.expand_path(File.dirname(__FILE__)+'/lib')
DATA_DIR = File.expand_path(File.dirname(__FILE__))+'/data'
DEBUG = true

APP_NAME					= 'Xmp Manager'
APP_COPYRIGHT			= "Copyright (c) #{Time.now.year} Luigi Maselli"
APP_AUTHORS 			= ['Luigi Maselli <luigix_@t_gmail.com>']
APP_VERSION				= 0.5
APP_LOGO_FILENAME = ''

GCONF_ROOT		= '/apps/xmp-manager'

require 'xmpmanager/selection'
require 'xmpmanager/ui/gtk'

XmpManager::MainWindow.new

#TODO: Manage errors w gui STDERR
